src/osd/osd_types.cc

```diff
@@ -1755,19 +1755,13 @@ void pg_pool_t::remove_snap(snapid_t s)
 {
   ceph_assert(snaps.count(s));
   snaps.erase(s);
-  snap_seq = snap_seq + 1;
 }
 
 void pg_pool_t::remove_unmanaged_snap(snapid_t s, bool preoctopus_compat)
 {
   ceph_assert(is_unmanaged_snaps_mode());
-  ++snap_seq;
   if (preoctopus_compat) {
     removed_snaps.insert(s);
-    // try to add in the new seq, just to try to keep the interval_set contiguous
-    if (!removed_snaps.contains(get_snap_seq())) {
-      removed_snaps.insert(get_snap_seq());
```

Comment:
It looks like the newly created snap got immediately removed -- why were we seeing holes?

Reply:
`pg_pool_t::removed_snaps`  was used (pre-octopus) only to track the non-pool removed snaps and to signify self-managed snaps pool.
To actually perform the removal `OSDMap::Incremental::new_removed_snaps` is used.
```
  case POOL_OP_DELETE_UNMANAGED_SNAP:
  ...
    pp.remove_unmanaged_snap(m->snapid, osdmap.require_osd_release < ceph_release_t::octopus);
    pending_inc.new_removed_snaps[m->pool].insert(m->snapid);
    changed = true;
```
```
  case POOL_OP_DELETE_SNAP:
    snapid_t s = pp.snap_exists(m->name.c_str());
    if (s) {
      pp.remove_snap(s);
      pending_inc.new_removed_snaps[m->pool].insert(s);
      changed = true;
    }
```

We no longer maintain `pg_pool_t::removed_snaps`, the flag `FLAG_SELFMANAGED_SNAPS` is used instead.
```
uint64_t pg_pool_t::add_unmanaged_snap(bool preoctopus_compat)
{
  if (snap_seq == 0) {
    if (preoctopus_compat) {
      // kludge for pre-mimic tracking of pool vs selfmanaged snaps.  after
      // mimic this field is not decoded but our flag is set; pre-mimic, we
      // have a non-empty removed_snaps to signifiy a non-pool-snaps pool.
      removed_snaps.insert(snapid_t(1));
    }
    snap_seq = 1;
  }
  flags |= FLAG_SELFMANAGED_SNAPS;
```